File: src/test/java/leetCode/simple/LeetCode20.java
package leetCode.simple;
import java.util.Stack;
/**
 * https://leetcode-cn.com/problems/valid-parentheses/
 *
 * @author chenwu on 2019.9.27
 */
public class LeetCode20 {

    private boolean checkCharIsPair(char left,char right){
        if(left == '(' && right == ')' ){
            return true;
        }
        if(left == '[' && right == ']'){
            return true;
        }
        if(left == '{' && right == '}'){
            return true;
        }
        return false;
    }

    public boolean isValid(String s) {
        if(s==null||s.length()==0){
            return true;
        }
        Stack<Character> stack = new Stack<Character>();
        for(int i=0;i<s.length();i++){
            char item = s.charAt(i);
            if(stack.isEmpty()){
                stack.push(item);
            }else{
                char topChar = stack.peek();
                if(checkCharIsPair(topChar,item)){
                    stack.pop();
                }else{
                    stack.push(item);
                }
            }
        }
        return stack.isEmpty();
    }

    public static void main(String[] args){
        LeetCode20 test = new LeetCode20();
        boolean result = test.isValid("{[]}");
        System.out.println(result);
    }
}

File: src/test/java/leetCode/simple/scala/Leetcode20.scala
package leetCode.simple.scala

import scala.collection.mutable.Stack

object Leetcode20 {

  def isValid(s: String): Boolean = {
      val stack = Stack[Char]()
      for(char<-s){
          if(stack.isEmpty){
              stack.push(char)
          }else{
              val topChar = stack.top
              if(((char == ')' && topChar=='(')) || ((char == ']' && topChar == '[')) || ((char=='}' && topChar == '{'))){
                  stack.pop()
              }else{
                  stack.push(char)
              }
          }
      }
      stack.isEmpty
  }

  def main(args: Array[String]): Unit = {
      val str = "{[]}"
      println(isValid(str))
  }
}
